torch/csrc/utils/schema_info.cpp

```diff
@@ -1,5 +1,83 @@
 #include <torch/csrc/utils/schema_info.h>
 
 namespace torch {
-namespace utils {} // namespace utils
+namespace utils {
+void SchemaInfo::addArgumentValue(
+    const std::string& name,
+    const at::IValue& value) {
+  c10::optional<int> index = schema_.argumentIndexWithName(name);
+  TORCH_INTERNAL_ASSERT(
+      index != c10::nullopt, "Schema has no argument named ", name);
+  value_map_[name] = value;
+  updated_ = false;
+}
+
+void SchemaInfo::addArgumentValues(
+    const std::vector<c10::optional<at::IValue>>& value_list) {
+  for (size_t i = 0; i < value_list.size(); i++) {
+    if (i < schema_.arguments().size() && value_list[i] != c10::nullopt) {
+      value_map_[schema_.arguments()[i].name()] = *(value_list[i]);
+      updated_ = false;
+    }
+  }
+}
+
+void SchemaInfo::addArgumentValues(
+    const std::unordered_map<std::string, at::IValue>& values) {
+  for (const auto& key_pair : values) {
+    addArgumentValue(key_pair.first, key_pair.second);
+  }
+}
+
+bool SchemaInfo::is_mutable() {
+  for (size_t i = 0; i < schema_.arguments().size(); i++) {
+    if (is_mutable(i)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool SchemaInfo::is_mutable(size_t index) {
+  TORCH_INTERNAL_ASSERT(
+      index < schema_.arguments().size(), "Invalid index for schema.");
+  if (!updated_) {
+    generateAliasMaps();
+  }
+  return std::any_of(
+      input_alias_map_[index].begin(),
+      input_alias_map_[index].end(),
+      [this](size_t index) { return this->schema_.is_mutable(index); });
+}
+
+bool SchemaInfo::is_mutable(c10::string_view name) {
+  c10::optional<int> index = schema_.argumentIndexWithName(name);
+  TORCH_INTERNAL_ASSERT(
+      index != c10::nullopt, "Schema has no argument named ", name);
+
+  return is_mutable(*index);
+}
+
+void SchemaInfo::generateAliasMaps() {
+  updated_ = true;
+  input_alias_map_ = std::vector<std::unordered_set<size_t>>(
+      schema_.arguments().size(), std::unordered_set<size_t>());
+  for (size_t i = 0; i < schema_.arguments().size(); i++) {
+    for (size_t j = i; j < schema_.arguments().size(); j++) {
```

Comment:
You should skip arguments that are not tensors.

Reply:
Is the reason for that for performance?
Also wouldn't we be worried about any possible container that contains a tensor (like List[List[Tensor]], Dict(Tensor) etc) in case future ops use these containers even if the current ops in native_functions.yaml do not? 
If that's the case, wouldn't the performance decrease from checking whether an IValue is an any level container around Tensor result in around the same level of performance as just considering every type? (I wouldn't be surprised if I'm missing some logic here about this)

Reply:
You're right, even with 10 arguments, the double for loop isn't that slow.